solidity/test/TestPhasedEscrow.js

```diff
@@ -426,7 +425,378 @@ describe("PhasedEscrow", () => {
+      beneficiaries = [
+        beneficiary1.address,
+        beneficiary2.address,
+        beneficiary3.address,
+      ]
+
+      amounts = [100, 200, 300]
+      escrowBalance = 600
+
+      await batchedPhasedEscrow.approveBeneficiary(beneficiary1.address, {
+        from: owner,
+      })
+      await batchedPhasedEscrow.approveBeneficiary(beneficiary2.address, {
+        from: owner,
+      })
+      await batchedPhasedEscrow.approveBeneficiary(beneficiary3.address, {
+        from: owner,
+      })
+
+      await batchedPhasedEscrow.transferDraweeRole(drawee, {
+        from: owner,
+      })
+
+      await token.transfer(batchedPhasedEscrow.address, escrowBalance, {
+        from: owner,
+      })
+    })
+
+    it("can be called by drawee", async () => {
+      await batchedPhasedEscrow.batchedWithdraw(beneficiaries, amounts, {
+        from: drawee,
+      })
+      // ok, no revert
+    })
+
+    it("can not be called by owner if not drawee", async () => {
+      await expectRevert(
+        batchedPhasedEscrow.batchedWithdraw(beneficiaries, amounts, {
+          from: owner,
+        }),
+        "Caller is not the drawee"
+      )
+    })
+
+    it("can not be called by non-drawee", async () => {
+      await expectRevert(
+        batchedPhasedEscrow.batchedWithdraw(beneficiaries, amounts, {
+          from: updatedDrawee,
+        }),
+        "Caller is not the drawee"
+      )
+    })
+
+    it("reverts when input arrays have different lengths", async () => {
+      await expectRevert(
+        batchedPhasedEscrow.batchedWithdraw(beneficiaries, [100, 200], {
+          from: drawee,
+        }),
+        "Mismatched arrays length"
+      )
+    })
+
+    it("reverts when beneficiary is not IBeneficiaryContract", async () => {
+      await expectRevert.unspecified(
+        batchedPhasedEscrow.batchedWithdraw(
+          [beneficiary1.address, beneficiary2.address, owner],
+          amounts,
+          {
+            from: owner,
+          }
+        )
+      )
+    })
+
+    it("reverts when beneficiary has not been approved", async () => {
+      const anotherBeneficiary = await TestSimpleBeneficiary.new({from: owner})
+
+      await expectRevert(
+        batchedPhasedEscrow.batchedWithdraw(
+          [beneficiary1.address, anotherBeneficiary.address], 
+          [100, 200], {
+          from: drawee,
+        }),
+        "Beneficiary not approved"
+      )
+    })
+
+    it("withdraws specified tokens to beneficiaries", async () => {
+      await batchedPhasedEscrow.batchedWithdraw(beneficiaries, amounts, {
+        from: drawee,
+      })
+
+      for (let i = 0; i < beneficiaries.length; i++) {
+        expect(await token.balanceOf(beneficiaries[i])).to.eq.BN(
+          amounts[i],
+          `Unexpected amount withdrawn for beneficiary ${i}`
+        )
+      }
+
+      expect(await token.balanceOf(batchedPhasedEscrow.address)).to.eq.BN(
+        0,
+        `Unexpected escrow balance`
+      )
+    })
```

Comment:
Can we add a test case where transaction reverts due to exceeding the available escrow balance?

Reply:
I was thinking about it but wasn't sure if that's the right thing to do. I do not want to add a custom `revert` because that would require adding a separate step to a loop summing up tokens withdrawn so far. The fact I can't spend more than I have should be enforced by ERC20 implementation and I do not want to duplicate this logic just to burn some gas. Given that, I can add a test but then I am testing ERC20 implementation: I can't spend more than I have.

Reply:
It's up to our contract how such situation is handled. We're using ERC20 impl for a single tansfer but our function performs multiple of them. We could handle revert of one transaction and let the others finish sucesfully. In the test we want to validate that if one transfer fails the whole call reverts.

Reply:
Good point, I will add a test.

Reply:
Added in the last commit, https://github.com/keep-network/keep-core/pull/2230/commits/686fbe387bc5e077839927265e2c32ad7c74214f.